### ISSUE TITLE
StochasticCost as a feature of layers

### ISSUE THREAD

**Issue by zaqqwerty** [tensorflow/quantum#411]
In the past, subsampling of PauliTerms from input PauliSums was taken care of by an option in the [SGDifferentiator module](https://github.com/tensorflow/quantum/blob/dd209265ea32d00fb9b38682b89b0556204b9ce0/tensorflow_quantum/python/differentiators/stochastic_differentiator.py#L38).  However, there were a few issues with its implementation there:

1. Instead of subsampling PauliTerms from PauliSums, it actually changed the shape of the measurement tensor by subsampling PauliSums.
2. Had a python implementation instead of a C++ op backend.
3. Does not fit the new differentiator interface being implemented in #409 .

Fortunately, in discussions on the design doc for the new differentiators, it looks like this could be a feature we still want to keep, and pull up the level of `tfq.layers`.  Since, this subsampling of PauliTerms can speed up the estimation of expectation values in any context, not just when seeking gradients.  See also #230 for the possible interaction of such a feature with Engine.  Thinking to take this on in the near future.  Thoughts on this @MichaelBroughton @jaeyoo ?

**Comment by jaeyoo** [comment]
@zaqqwerty I agree with that. The first target application was the SGDifferentiator, but StochasticCost or StochasticLoss itself is a standalone. Thank you for your research!

**Comment by MichaelBroughton** [comment]
Perhaps there is a nice way we could work this in throughout the entire stack in TFQ. Maybe we can define an op flag for the op returned by `get_sampled_expectation_op` that has `bool: subsample_paulis=True/False` (https://www.tensorflow.org/guide/create_op#attrs) From there it would be simple to have the keras layers choose values for this based on user preferences. What do people think ?
